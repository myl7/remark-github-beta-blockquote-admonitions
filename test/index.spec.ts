// Copyright (C) myl7
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'
import { parseDocument } from 'htmlparser2'
import { selectOne, selectAll } from 'css-select'
import { remark } from 'remark'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'
import plugin, { Config, mkdocsConfig } from '../src/index.js'

function defineCase(
  name: string,
  options: {
    input: string
    assertions: (html: string) => Promise<void> | void
    config?: Partial<Config>
  },
) {
  it(name, async function () {
    const processor = remark()
      .use(remarkParse)
      // To make unified happy, since it does not allow (for types) to pass an undefined as config.
      // This test code can also handle the difference between `.use(plugin)` and `.use(plugin, {})`.
      .use(plugin, ...(options.config ? [options.config] : []))
      .use(remarkRehype)
      .use(rehypeStringify)

    const html = String(await processor.process(options.input))
    await options.assertions(html)

    expect(html).toMatchSnapshot()
  })
}

describe('GitHub beta blockquote-based admonitions with titles like [!NOTE]', function () {
  defineCase('should transform', {
    input: `\
# Admonitions
> [!NOTE]
> test
`,
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should not transform when single line', {
    input: `\
# Admonitions
> [!NOTE] test
`,
    assertions(html) {
      const elem = selectOne('div.admonition', parseDocument(html))
      expect(elem).to.have.be.null
    },
  })

  defineCase('should transform with nested ones', {
    input: `\
# Admonitions
> [!NOTE]
> test
>
> > [!NOTE]
> > test
> >
> > > [!WARNING]
> > > test
`,
    assertions(html) {
      const elem = selectOne(
        'div.admonition > div.admonition > div.admonition > p.admonition-title:first-child',
        parseDocument(html),
      )
      expect(elem).to.have.nested.property('firstChild.data', 'WARNING')
    },
  })

  defineCase('should transform with unordered lists from issue #4', {
    input: `\
# Admonitions
> [!NOTE]
> - Here you go
> - Here you go again
> - Here you go one more time
`,
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should transform with ordered lists from issue #4', {
    input: `\
# Admonitions
> [!NOTE]
> 1. Here you go
> 2. Here you go again
> 3. Here you go one more time
`,
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should not transform when title is not in form [!NOTE] but legacy **Note**', {
    input: `\
# Admonitions
> **Note**
> test
`,
    assertions(html) {
      const elem = selectOne('div.admonition', parseDocument(html))
      expect(elem).to.be.null
    },
  })

  defineCase('should transform with title with trailing whitespaces to be trimmed', {
    input: `\
# Admonitions
> [!NOTE] \r\t\v\

> test
`,
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should transform GitHub example (until 2024-01-02) with default config', {
    input: `\
# Admonitions
> [!NOTE]
> Highlights information that users should take into account, even when skimming.

> [!TIP]
> Optional information to help a user be more successful.

> [!IMPORTANT]
> Crucial information necessary for users to succeed.

> [!WARNING]
> Critical content demanding immediate user attention due to potential risks.

> [!CAUTION]
> Negative potential consequences of an action.
`,
    assertions(html) {
      const elems = selectAll('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elems).to.have.lengthOf(5)
      expect(elems[0]).to.have.nested.property('firstChild.data', 'NOTE')
      expect(elems[1]).to.have.nested.property('firstChild.data', 'TIP')
      expect(elems[2]).to.have.nested.property('firstChild.data', 'IMPORTANT')
      expect(elems[3]).to.have.nested.property('firstChild.data', 'WARNING')
      expect(elems[4]).to.have.nested.property('firstChild.data', 'CAUTION')
    },
  })

  defineCase('should transform when title has 2 trailing spaces from issue #12', {
    input:
      '# Admonitions  \n' +
      '> [!WARNING]\n' +
      '> Critical content demanding immediate user attention due to potential risks.\n',
    config: {
      classNameMaps: {
        block: (title: string) => `admonition admonition--${title.toLowerCase()}`,
        title: (title: string) => `admonition-title admonition-title--${title.toLowerCase()}`,
      },

      titleTextMap: (title: string) => {
        // Removes `![]`
        let displayTitle = title.slice(2, -1)
        // Capitalize
        displayTitle = `${displayTitle.at(0)}${displayTitle.substring(1).toLowerCase()}`
        return {
          displayTitle,
          checkedTitle: displayTitle,
        }
      },
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'Warning')
    },
  })
})

describe('the plugin options for titles like [!NOTE]', function () {
  defineCase('should accept custom class names', {
    input: `\
# Admonitions
> [!NOTE]
> test
`,
    config: {
      classNameMaps: {
        block: 'ad',
        title: ['ad-title1', 'ad-title2'],
      },
    },
    assertions(html) {
      const elem = selectOne('div.ad > p.ad-title1.ad-title2:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should accept custom class names with functions', {
    input: `\
# Admonitions
> [!NOTE]
> test
`,
    config: {
      classNameMaps: {
        block: (title) => `ad-${title.toLowerCase()}`,
        title: (title) => [`ad-${title.toLowerCase()}-title1`, `ad-${title.toLowerCase()}-title2`],
      },
    },
    assertions(html) {
      const elem = selectOne('div.ad-note > p.ad-note-title1.ad-note-title2:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should accept custom title filter', {
    input: `\
# Admonitions
> [!TIPS]
> test
`,
    config: {
      titleFilter: ['[!TIPS]', '[!HINTS]'],
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'TIPS')
    },
  })

  defineCase('should accept custom title filter with functions', {
    input: `\
# Admonitions
> [!tIps]
> test
`,
    config: {
      titleFilter: (title) => title.startsWith('[!tIp') && title.endsWith(']'),
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'tIps')
    },
  })

  defineCase('should accept title text map to customize title text', {
    input: `\
# Admonitions
> [!NOTE:OKOK]
> test
`,
    config: {
      titleFilter: (title) => title.startsWith('[!NOTE:') && title.endsWith(']'),
      titleTextMap: (title) => {
        const titleSplit = title.substring(2, title.length - 1).split(':')
        return { displayTitle: titleSplit[1], checkedTitle: titleSplit[0] }
      },
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'OKOK')
    },
  })

  defineCase('should accept title text map to customize title text with spaces', {
    input: `\
# Admonitions
> [!NOTE: OK OK ]
> test
`,
    config: {
      titleFilter: (title) => title.startsWith('[!NOTE:') && title.endsWith(']'),
      titleTextMap: (title) => {
        const titleSplit = title.substring(2, title.length - 1).split(':')
        return { displayTitle: titleSplit[1], checkedTitle: titleSplit[0] }
      },
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', ' OK OK ')
    },
  })

  defineCase('should accept data maps to edit data', {
    input: `\
# Admonitions
> [!NOTE]
> test
`,
    config: {
      dataMaps: {
        block: (data) => ({ ...data, hName: 'admonition' }),
        title: (data) => data,
      },
    },
    assertions(html) {
      const elem = selectOne('admonition.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })

  defineCase('should accept title with trailing whitespaces with custom whitespace handling', {
    input: `\
# Admonitions
> [!NOTE] \t\v\

> test
`,
    config: {
      titleFilter: (title) => title.startsWith('[!NOTE]'),
      titleTextMap: (title) => ({
        displayTitle: title.substring(2, title.length - 1 - 3),
        checkedTitle: title.substring(2, title.length - 1 - 3),
      }),
      titleKeepTrailingWhitespaces: true,
    },
    assertions(html) {
      const elem = selectOne('div.admonition > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', 'NOTE')
    },
  })
})

describe('MkDocs admonition HTML options for titles like [!NOTE]', function () {
  defineCase('should transform', {
    input: `\
# Admonitions
> [!note danger "Don't try this at home"]
> You should note that the title will be automatically capitalized.
`,
    config: mkdocsConfig,
    assertions(html) {
      const elem = selectOne('div.admonition.note.danger > p.admonition-title:first-child', parseDocument(html))
      expect(elem).to.have.nested.property('firstChild.data', "Don't try this at home")
    },
  })

  defineCase('should transform with custom types', {
    input: `\
# Admonitions
> [!admonition: guess "Don't try this at home"]
> You should note that the title will be automatically capitalized.
`,
    config: mkdocsConfig,
    assertions(html) {
      const doc = parseDocument(html)
      const elem = selectOne('div.admonition.guess > p.admonition-title:first-child', doc)
      expect(elem).to.have.nested.property('firstChild.data', "Don't try this at home")
      const elemUnexpected = selectOne('div.admonition.admonition\\:', doc)
      expect(elemUnexpected).to.be.null
    },
  })
})
