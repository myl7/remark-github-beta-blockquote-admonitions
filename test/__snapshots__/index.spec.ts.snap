// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should not transform when single line 1`] = `
"<h1>Admonitions</h1>
<blockquote>
<p>[!NOTE] test</p>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should not transform when title is not in form [!NOTE] but legacy **Note** 1`] = `
"<h1>Admonitions</h1>
<blockquote>
<p><strong>Note</strong>
test</p>
</blockquote>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform GitHub example (until 2024-01-02) with default config 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>Highlights information that users should take into account, even when skimming.</p>
</div>
<div class="admonition">
<p class="admonition-title">TIP</p>
<p>Optional information to help a user be more successful.</p>
</div>
<div class="admonition">
<p class="admonition-title">IMPORTANT</p>
<p>Crucial information necessary for users to succeed.</p>
</div>
<div class="admonition">
<p class="admonition-title">WARNING</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>
<div class="admonition">
<p class="admonition-title">CAUTION</p>
<p>Negative potential consequences of an action.</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform when admonition title has 2 trailing spaces (carriage return) from issue #12 1`] = `
"<h1>Admonitions</h1>
<div class="admonition admonition--warning">
<p class="admonition-title admonition-title--warning">Warning</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform when title has 2 trailing spaces from issue #12 1`] = `
"<h1>Admonitions</h1>
<div class="admonition admonition--warning">
<p class="admonition-title admonition-title--warning">Warning</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with nested ones 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
<div class="admonition">
<p class="admonition-title">WARNING</p>
<p>test</p>
</div>
</div>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with ordered lists from issue #4 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p></p>
<ol>
<li>Here you go</li>
<li>Here you go again</li>
<li>Here you go one more time</li>
</ol>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with title with trailing whitespaces to be trimmed 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
</div>"
`;

exports[`GitHub beta blockquote-based admonitions with titles like [!NOTE] > should transform with unordered lists from issue #4 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p></p>
<ul>
<li>Here you go</li>
<li>Here you go again</li>
<li>Here you go one more time</li>
</ul>
</div>"
`;

exports[`MkDocs admonition HTML options for titles like [!NOTE] > should transform 1`] = `
"<h1>Admonitions</h1>
<div class="admonition note danger">
<p class="admonition-title">Don't try this at home</p>
<p>You should note that the title will be automatically capitalized.</p>
</div>"
`;

exports[`MkDocs admonition HTML options for titles like [!NOTE] > should transform with custom types 1`] = `
"<h1>Admonitions</h1>
<div class="admonition guess">
<p class="admonition-title">Don't try this at home</p>
<p>You should note that the title will be automatically capitalized.</p>
</div>"
`;

exports[`should transform when admonition title has 2 trailing spaces (carriage return) from issue #12 1`] = `
"<h1>Admonitions</h1>
<div class="admonition admonition--warning">
<p class="admonition-title admonition-title--warning">Warning</p>
<p>Critical content demanding immediate user attention due to potential risks.</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept custom class names 1`] = `
"<h1>Admonitions</h1>
<div class="ad">
<p class="ad-title1 ad-title2">NOTE</p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept custom class names with functions 1`] = `
"<h1>Admonitions</h1>
<div class="ad-note">
<p class="ad-note-title1 ad-note-title2">NOTE</p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept custom title filter 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">TIPS</p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept custom title filter with functions 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">tIps</p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept data maps to edit data 1`] = `
"<h1>Admonitions</h1>
<admonition class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
</admonition>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept title text map to customize title text 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">OKOK</p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept title text map to customize title text with spaces 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title"> OK OK </p>
<p>test</p>
</div>"
`;

exports[`the plugin options for titles like [!NOTE] > should accept title with trailing whitespaces with custom whitespace handling 1`] = `
"<h1>Admonitions</h1>
<div class="admonition">
<p class="admonition-title">NOTE</p>
<p>test</p>
</div>"
`;
